keps/sig-node/20191205-container-streaming-requests.md

```diff
@@ -0,0 +1,211 @@
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+There are several features and flags for handling container streaming requests
+(exec/attach/port-forward). Having so many different ways of handling these requests unnecessarily
+complicates configuration, and opens the cluster to potential security risks. This is a proposal to
+consolidate the current options into a single supported configuration. Specifically, it proposes:
+
+1. Deprecate the Kubelet's `--redirect-container-streaming` feature
+2. Deprecate the `StreamingProxyRedirects` feature
+
+## Motivation
+
+The `StreamingProxyRedirects` feature was originally added as part of the migration the [Container
+Runtime Interface][]. One of the initial goals was to bypass the Kubelet for these long-running
+requests in order to better account the request overhead. However, the apiserver is still
+responsible for proxying these requests, so the gains of bypassing the Kubelet are
+marginal. Furthermore, the redirect handling significantly complicates the apiserver stream
+handling, and opens it up to [potential SSRF attacks][], so we added the `ValidateProxyRedirects`
+feature to mitigate those.
+
+When we went to add mTLS support to the runtime streaming server, we realized getting the right
+certs established between the apiserver and the runtime was complicated, but the apiserver already
+had an mTLS connection to the Kubelet. So, we added the ability for the Kubelet to connect the
+stream locally and proxy the connection back to the apiserver. This setup is more secure, removes
+the need for `StreamingProxyRedirects` and `ValidateProxyRedirects`, but adds the Kubelet back into
+the long-running request chain.
+
+`StreamingProxyRedirects` and `ValidateProxyRedirects` are currently in Beta, and should either be
+graduated or removed.
+
+[Container Runtime Interface]: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/container-runtime-interface-v1.md
+[potential SSRF attacks]: https://github.com/kubernetes/kubernetes/issues/85867
+
+### Goals
+
+1. Simplify secure container runtime setup and operation
+2. Remove conformance dependencies on beta features
+
+### Non-Goals
+
+Design a high-bandwidth low-latency production ready streaming channel to containers. Such a system
+should go through an ingress point other than the apiserver, and could be implemented (or at least
+prototyped) out of tree by leveraging the new [ephemeral containers
+feature](20190212-ephemeral-containers.md).
+
+## Proposal
+
+After the [transition period](#rollout-plan), the only way to configure container streaming will be
```

Comment:
This is a bit confusing to follow as the --redirect-container-streaming flag is also being deprecated and removed.

Can we add a section that documents how the only way to configure container streaming will work upon completion?  As written, we will be referencing this through 1.22 so it would help to state how the feature in the end continues to work...

Reply:
Done.